### PR TITLE
onnx-systemds implementation

### PR DESCRIPTION
**Pull request by lukas-jkl**
This PR implements a first poc-implementation for an ONNX importer.
It adds support for the following operators:
* Add
* Sub
* MatMul
* Neg
* Xor
* Or
* And
* Relu
* Tanh
* Sigmoid
* Softmax
* Dropout
* MaxPool
* Conv
* If

As well as the logic for nested sub-graphs.

### CODE REVIEW

**Comment by j143** [comment]
I am reviewing this. Should be back in a day. thank you.

**Comment by lukas-jkl** [comment]
> On a cursory check, the changeset looks good. Can the design be documented thoroughly in a `.md` file.
> Feel free to include as much information in the doc as possible with the uses of tables, graphs etc.
> 
> I am facing problems with building `onnx` for now.

Thank you for the quick review, I've added a design document which describes the current implementation in detail. 
I think the conda install of onnx works best, I also had problems installing with other methods.

**Comment by j143** [comment]
Thank you for the design documentation, @lukas-jkl . I would make a deep pass shortly.

**Comment by j143** [comment]
Sorry, @lukas-jkl - I may not be able to review PR anytime soon. But, I posted this on the list, so that someone would come forward for review. Thank you.

**Comment by mboehm7** [comment]
thanks for the initial look @j143. @Baunsgaard could you please have a look. I'll also try it out later this week.

File: src/main/python/onnx_systemds/operator_gen.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+        function_namespace=function_namespace,
+        function=function_name,
+        arguments=[X, p, seed],
+        outputs=list(node.output),
+        doc_string=node.doc_string
+    )
+    return GeneratedScriptPart(imports=[import_render], dml_script=node_render)
+
+
+def gen_maxpool_call(env: jinja2.environment.Environment, graph: onnx.GraphProto,
```
Comment (Baunsgaard) [inline src/main/python/onnx_systemds/operator_gen.py:188]: gen conv and gen maxpool, contain much of the same code, maybe some of it could be alleviated using some private helper functions?

File: src/main/python/onnx_systemds/tests/util.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    :return: the return-code of systemds
+    """
+    if args is None:
+        args = []
+
+    systemds_root_path = resolve_systemds_root()
+
+    try:
+        abspath_input = os.path.abspath(input_file)
+        res = subprocess.run([systemds_root_path + "/bin/systemds.sh", abspath_input] + args,
```
Comment (Baunsgaard) [inline src/main/python/onnx_systemds/tests/util.py:40]: This is the issue for why it fails now. we changed the name to `/bin/systemds` without the `.sh`.

**Comment by Baunsgaard** [comment]
> > On a cursory check, the changeset looks good. Can the design be documented thoroughly in a `.md` file.
> > Feel free to include as much information in the doc as possible with the uses of tables, graphs etc.
> > I am facing problems with building `onnx` for now.
> 
> Thank you for the quick review, I've added a design document which describes the current implementation in detail.
> I think the conda install of onnx works best, I also had problems installing with other methods.

i installed not using conda, with no problems. the only thing to remember is to install protobuf, but the Onnx guide specifies this as well.

**Comment by Baunsgaard** [comment]
The thing to discuss in general for this implementation of Onnx in SystemDS is the use of jinja2. 
I would argue that it is fine, but it does come with some drawbacks.

1. Extra dependency 2. New syntax to learn.

The upside is somewhat hopefully simpler translation to DML. This said in the current implementation it would not be so difficult to simplify and construct the strings needed without jinja2, at least to my understanding, but maybe it is useful down the road.

Any opinions?

**Comment by lukas-jkl** [comment]
Thank you for the quick review @Baunsgaard ! I've implemented your suggested changes and have rewritten the documentation in rst files. I am however still a bit unsure about the yml files. I guess it would not be sufficient to simply add onnx and Jinja to the pip dependencies as protobuf would also need to be installed?

File: src/main/python/onnx_systemds/onnx_helper.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                inputs_with_initializers.append((input, initializer))
+
+    return inputs_with_initializers
+
+
+class PreparedValue:
+    """ Class for preparing onnx value structures for writing them to the dml script """
+    def __init__(self, value_info: onnx.ValueInfoProto, initializer: onnx.TensorProto = None):
+
+        supported_types = ["int", "boolean", "double"]
```
[the remark this answers is not in the document]
Comment (lukas-jkl) [reply src/main/python/onnx_systemds/onnx_helper.py]: Those should be the types supported by systems (I've now renamed the variable to make this more clear). So I guess it should be "integer"

File: src/main/python/onnx_systemds/onnx_helper.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            5: "int",  # int16_t
+            6: "int",  # int32_t
+            7: "int",  # int64_t
+            8: "string",
+            9: "boolean",  # bool
+
+            10: "double",  # float16,
+            11: "double",
+            12: "int",  # uint32
+            13: "int",  # uint64
```
[the remark this answers is not in the document]
Comment (lukas-jkl) [reply src/main/python/onnx_systemds/onnx_helper.py]: Yes, you are correct, I am however unsure how big the "Integer" type of systemds is? I guess it should be int32. I've now changed those translations and an exception is thrown if we do not support the type. I've also added an additional check to enforce that a matrix must only contain double values, as stated in the dml-language-reference.

File: src/main/python/onnx_systemds/convert.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    :param input_onnx_file: the onnx input file
+    :param output_dml_file: (optional) the dml output file,
+        if this parameter is not given the output file will have the same name as the input file
+    """
+    if not os.path.isfile(input_onnx_file):
+        raise Exception("Invalid input-file: " + str(input_onnx_file))
+
+    if not output_dml_file:
+        output_dml_file = os.path.splitext(ntpath.basename(input_onnx_file))[0] + ".dml"
```
[the remark this answers is not in the document]
Comment (lukas-jkl) [reply src/main/python/onnx_systemds/convert.py]: I guess os.path can also be used. I've now removed ntpath.

**Comment by lukas-jkl** [comment]
> > Thank you for the quick review @Baunsgaard ! I've implemented your suggested changes and have rewritten the documentation in rst files. I am however still a bit unsure about the yml files. I guess it would not be sufficient to simply add onnx and Jinja to the pip dependencies as protobuf would also need to be installed?
> 
> Since our python tests are executed on Ubuntu-latest it should **hopefully** be as simple as adding a step,
> 
> ```yaml
> - name: install protobuf
>    run: {INSERT INSTALL COMMAND HERE }
> ```

Seems to work now!

**Comment by Baunsgaard** [comment]
Cool many thanks, LGTM. :1st_place_medal: 
(if i had the power i would merge it)

Up for discussion in the future is still https://github.com/apache/systemml/pull/904#issuecomment-627933153

Only real thing left to refine in my opinion is the Output files for validation, that i think is to much, because the tests should not reflect a specific __result__, but a certain __behavior__. But I also acknowledge this is a highly personal thing, and that some of the python tests already in does not follow this principle.

@mboehm7

**Comment by mboehm7** [comment]
LGTM - thanks @lukas-jkl for this awesome patch, and @Baunsgaard for the detailed discussions. Overall, this is a great initial importer and completes the AMLS project. During the merge, I only added `**/*.out` to the exclude list of the rat check (for proper licenses).

Down the the road, I would also prefer to remove these expected output files and rather generate them dynamically (e.g., via pytorch). Similarly, we should clean up the license headers to use a common text layout (beginning/end lines).   

Btw, there are additional rat problems in the data slicing implementation (which I fix when merging Svetlana's PR) and the `JolEstimate*` tests (@Baunsgaard you might want to fix that).
